lib/kubeclient/common.rb

```diff
@@ -508,5 +569,11 @@ def http_options(uri)
 
       options.merge(@socket_options)
     end
+
+    def http_err(json_error_msg, e)
+      err_message = json_error_msg['message'] || e.message
+      error_klass = e.http_code == 404 ? ResourceNotFoundError : HttpError
+      raise error_klass.new((e.http_code || 404), err_message, e.response)
```

Comment:
why did you add the `|| 404`?
I guess http_code can be missing if error happened earlier (socket refused etc), in which case setting 404 is misleading.

Reply:
for the synchrounous `get_entities`, we use `RestClient::Resource.new().get` which returns `ResourceNotFoundError` if the HTTP status code is a 404.
for the async `get_entities`, we use `RestClient::Request.execute()` which does **not** return an error on 404. the purpose of this change is to catch the 404 and propagate a `ResourceNotFoundError` (to maintain consistency with the public API of this function)

Reply:
Hmm, perhaps I see now.
The scenario you talk about is `handle_exceptions` rescuing the `raise ResourceNotFoundError` within `block` on L333?
And that's Kubeclient's ResourceNotFoundError and it has no `.http_code`?  If so, how about raising `RestClient::NotFound` there, with http_code == 404, and letting this code translate it to `raise ResourceNotFoundError`?

Reply:
`handle_exception` tries to parse `e.response` assuming it's a string. `RestClient::NotFound.response` is the HTTP response object. Unless we change the api of that method, raising our own `RestClient::NotFound` causes the error:

```
Message: <"no implicit conversion of Net::HTTPNotFound into String">
---Backtrace---
/home/ilackarms/.rbenv/versions/2.3.0/lib/ruby/2.3.0/json/common.rb:156:in `initialize'
/home/ilackarms/.rbenv/versions/2.3.0/lib/ruby/2.3.0/json/common.rb:156:in `new'
/home/ilackarms/.rbenv/versions/2.3.0/lib/ruby/2.3.0/json/common.rb:156:in `parse'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:114:in `rescue in handle_exception'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:111:in `handle_exception'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:516:in `get_entities_async'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:259:in `get_entities'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:180:in `block (2 levels) in define_entity_methods'
/home/ilackarms/workspace/manageiq/kubeclient/lib/kubeclient/common.rb:91:in `method_missing'
/home/ilackarms/workspace/manageiq/kubeclient/test/test_kubeclient.rb:242:in `block in test_entity_list_with_block_404'
```